src/Headers/NeighbourManager.h

```diff
@@ -0,0 +1,321 @@
+  vector<ParticleType > neibdata;
+
+  vector<int> culled_neiblist;
+  vector<int> smoothgravlist;
+
+  vector<FLOAT> dr;
+  vector<FLOAT> drmag;
+
+  const DomainBox<ndim>* _domain;
+  const ParticleTypeRegister* _types ;
+  double _kernrange ;
+
+  typedef struct {}        _true_type ;
+  typedef struct {char c;} _false_type ;
+
+public:
+  using NeighbourManagerDim<ndim>::tempneib;
+  using NeighbourManagerDim<ndim>::tempperneib;
+  using NeighbourManagerDim<ndim>::tempdirectneib;
+  using NeighbourManagerDim<ndim>::gravcell;
+
+  NeighbourManager(const Hydrodynamics<ndim>* hydro, const DomainBox<ndim>& domain)
+  : _NPeriodicGhosts(0), _NCellDirectNeib(0),
+    _domain(&domain),
+    _types(&(hydro->types)),
+    _kernrange(hydro->kernp->kernrange)
+  { } ;
+
+  NeighbourManager(const ParticleTypeRegister& types, double kernrange,
+      const DomainBox<ndim>& domain)
+  : _NPeriodicGhosts(0), _NCellDirectNeib(0),
+    _domain(&domain),
+    _types(&types),
+    _kernrange(kernrange)
+  { } ;
+
+  void clear() {
+    NeighbourManagerDim<ndim>::clear() ;
+    neiblist.clear();
+    neibdata.clear();
+    neib_idx.clear();
+    directlist.clear();
+  }
+
+
+  template<class InParticleType>
+  void EndSearch(const TreeCellBase<ndim> &cell, const InParticleType* partdata) {
+    _EndSearch(cell, partdata, false) ;
+  }
+
+  template<class InParticleType>
+  void EndSearchGravity(const TreeCellBase<ndim> &cell, const InParticleType* partdata) {
+    _EndSearch(cell, partdata, true) ;
+  }
+
+  int GetNumAllNeib() {
+    return neib_idx.size();
+  }
+
+  std::pair<int,ParticleType*> GetNeibI(int i) {
+    return make_pair(neib_idx[i],&neibdata[i]);
+  }
+
+  template<class InParticleType>
+  int GetParticleNeib(const InParticleType& p,const Typemask& hydromask, int** neiblist_p,
+      ParticleType** neibdata_p, const bool do_pair_once) {
+
+
+    if (do_pair_once)
+      TrimNeighbourLists<InParticleType,_true_type>(p, hydromask, false) ;
+    else
+      TrimNeighbourLists<InParticleType,_false_type>(p, hydromask, false) ;
+
+
+    *neiblist_p=&culled_neiblist[0];
+    *neibdata_p=&neibdata[0];
+    return culled_neiblist.size();
+
+  }
+
+  template<class InParticleType>
+  ListLength GetParticleNeibGravity(const InParticleType& p,const Typemask& hydromask, int** neiblist_p,
+      int** directlist_p, int** smoothgravlist_p, ParticleType** neibdata_p) {
+
+    TrimNeighbourLists<InParticleType,_false_type>(p, hydromask, true) ;
+
+    ListLength listlength;
+    listlength.Nhydro=culled_neiblist.size();
+    listlength.Ndirect=directlist.size();
+    listlength.Ngrav=smoothgravlist.size();
+
+    *neiblist_p = &neiblist[0];
+    *directlist_p = &directlist[0];
+    *smoothgravlist_p = &smoothgravlist[0];
+    *neibdata_p = &neibdata[0];
+
+    return listlength;
+
+  }
+
+private:
+
+  template<class InParticleType>
+   void _EndSearch(const TreeCellBase<ndim> &cell, const InParticleType* partdata, bool keep_direct=true) {
```

Comment:
What exactly does this function do?  Its name suggests it's maybe some last step or deallocation/termination but actually it does a hell of a lot of work and I'm only able to determine that by reading through the whole function.  A lot of these new neighbour classes are perhaps not sufficiently commented, which is fine for most of it since it's quite well-written 'self-commenting' code.  But function names should be clear precisely what they are doing OR have a few sentences explaining what they do in a comment-block (and preferably both).